Strings Library (#12)
FuelLabs/sway-libs: sway_libs/src/string/SPECIFICATION.md

> # Overview

Reviewer: Here is a list of functions/traits I would like to see implemented in the String library that currently are missing.

`push_str`
`reserve`
`String::from(String | str)`

Author: `push_str` is not currently possible. A `str` of two different lengths are treated as seperate types and I would not be able to iterate over a `str`.

`reserve` should be implemented at the `vec` level, and then used here.

`String::from()` again has the same problem as `push_str`

## Public Functions

### `as_bytes`

The `as_bytes` function is used to convert the `String` struct back to a `Vec` of `u8` bytes. 

### `capacity`

The `capacity` function will return the current amount of memory on the heap allocated to the `String`. 

> **Note** Capacity and length are not the same. A `String` may have a length of 0 but any arbitrary capacity.

### `clear`

The `clear` function will truncate the `String` to a length of 0. This does not clear the capacity of the `String`.

### `from_utf8`

The `from_utf8` function will return a `String` from a `Vec` of `u8` bytes.

### `insert`

The `insert` function will insert a new element at the specified index in the `String`. 

### `is_empty`

The `is_emtpy` function will return a `bool` as to whether there are elements in the `String`. This does not check the capacity of the `String`.

### `len`

The `len` function will return the length or number of elements in the `String`. This does not check the capacity of the `String`.

### `new`

The `new` function will construct an empty `String` struct.

### `nth`

> The `nth` function will return the element at the specified index in the `String`. If the index is out of bounds, `None` is returned.

Author: No, it does not. Because Rust's strings are UTF-8 encoded, the concept of indexing itself would be ambiguous. If a single character is 4 bytes, would this be `string[0]` or `string[3]`?

I will update the wording to mention this.

# Overview

This document provides an overview of the String library.

It outlines the use cases, i.e. specification, and describes how to implement the library.

## Use Cases

> The String library can be used anytime a string's length is unknown at compile time and as such lives on the heap. Further methods can then be implemented to provide additional features building off of the `String` struct.

Reviewer 2: ```suggestion
The String library can be used anytime a string's length is unknown at compile time. Further methods can then be implemented to provide additional features building off of the `String` struct.
```

Irrelevant for a use case description. The user only cares about having a dynamic string rather than the static built-in. The implementation can be mentioned in some technical section somewhere if it's relevant

> **Note** There is no guarantee in the validity of the UTF-8 encoded `String` and should be used with caution. For more information, please see the [known issues](./README.md#known-issues).

## Public Functions

### `as_bytes()`

> Used to convert the `String` struct to a `Vec` of `u8` bytes. 

Reviewer 2: There is no conversion here. It returns the wrapped value which is the vec

### `capacity()`

Returns the total amount of memory on the heap allocated to the `String` which can be filled with bytes. 

> > **Note** Capacity and length are not the same. A `String` may have a length of 0 but any arbitrary capacity.

Reviewer 2: Positive capacity rather than arbitrary

### `clear()`

Truncates the `String` to a length of 0 and will appear empty. This does not clear the capacity of the `String`.

### `from_utf8()`

> Given a vector of `u8`'s a new `String` instance will be returned.

Reviewer 2: ```suggestion
A new instance of a `String` will be created from a vector of `u8`'s
```

### `insert()`

Inserts a new byte at the specified index in the `String`. 

### `is_empty()`

Returns a boolean indicating whether the length of the `String` is zero.

### `len()`

Returns the total number of bytes in the `String`. 

### `new()`

Creates a new instance of the `String` struct.

### `nth()`

Returns the byte at the specified index in the `String`. If the index is out of bounds, `None` is returned.

### `pop()`

Removes the last byte in the `String` and returns it. If the `String` does not have any bytes, `None` is returned. 

### `remove()`

> Will both remove and return the specified byte in the `String`. 

Reviewer 2: ```suggestion
Remove and return the specified byte in the `String`. 
```
